pelican/plugins/pelican_redirect/__init__.py

```diff
@@ -28,10 +28,13 @@
 """
 
 
-@dataclass
 class Redirect:
-    to: contents.Content
-    from_url: str
+    def __init__(self, to: contents.Content, from_url: str):
+        self.to = to
+        if from_url[-1] == '/' or os.path.splitext(from_url)[-1] == '':
+            self.from_url = from_url + ('' if from_url[-1] == '/' else '/') + 'index.html'
+        else:
+            self.from_url = from_url
```

Comment:
I think this can still be a `dataclass` and it can use the [`__post_init__` method](https://docs.python.org/3/library/dataclasses.html#post-init-processing):

```python
def __post_init__(self):
    if not self.from_url.endswith(".html"):
        self.from_url = self.from_url.rstrip("/") + "/index.html"
```
I'm pretty sure that'll still work the way you're suggesting 😄

Reply:
Cool, I'm happy to keep it a dataclass. I didn't know about `__post_init__`.

I'm just thinking that it we check explicitly for the `.html` extension, it is going to miss `.htm`, as well as other extensions such as `.php` and `.aspx`. Someone migrating from a blog platform using PHP or ASP.NET would have such file extensions. Assuming someone is migrating to a static site generator for simplicity's sake, they probably won't have PHP or ASP.NET installed on their web server, so they can simply configure their web server to serve `.php` and `.aspx` as HTML.

That said, your suggestion might work (I guess depending on the specific web server) if someone for some reason wants to run their static site on a server which does have PHP etc. installed on it.

Reply:
Good point, I didn't think of other extensions. I think it'd be reasonable to add `.htm` into a tuple inside the `endswith` to check for that as well.

For dynamic frameworks that require PHP or .NET or other languages (even Python), I don't think there's a way to support anything other than a directory approach. That said, I haven't used WordPress in _years_ and not used any other frameworks in other languages, so I might be misunderstanding.

For instance, if you had a post at `http://example.com/blog/post-1.php`, Pelican can't write out PHP files, but you could sort of fake it by adding `/index.html` on the end, which this will do as-is. Similarly, if you have `...blog/posts.php?post=post_1` as your original URL, a simple static HTML file isn't going to be able to handle that without help from the web server, at which point you might as well just redirect using the web server.

Let me know what you think!

Reply:
Sure, yes, sorry, I didn't mean to imply that Pelican could generate PHP code. What I meant is that languages such as PHP generate HTML files which are served with the `.php` extension. But if one removes the PHP engine from the equation, and tell your web server to treat files ending in `.php` as plain old HTML files, it would read the HTTP redirect header and go to the Pelican generated content.

For example, let's say I have `my-legacy-article.md` which should redirect from `my-first-post.php`. Pelican generates `my-legacy-article.html` and `my-first-post.php`. Because I run a static server and don't have PHP installed on it, I define in the web server that the MIME type `text/html` should be any file which ends in `htm`, `html`, `shtml`, `php`. When a user clicks a link which still points to `my-first-post.php`, the web server serves the content of `my-first-post.php` as-is as HTML. The web server reads the HTTP, and gets redirected to `my-legacy-article.html`.

Admittedly, this is a hacky solution, because it requires telling the web server to treat `.php` files as HTML files.

If web servers won't get confused serving directories ending in `.php`, then your solution is cleaner. I would hope that web servers are smart enough to not get confused, but I don't know the actual behaviour, and haven't tested it yet.  In other words, if I were to have PHP installed on my server, and I have the files `a.php` and `b.php/index.html`, would the web server be smart enough to know that `http://example.com/a.php` should serve `a.php` through PHP, and `http://example.com/b.php` is a directory and it should server `b.php/index.html`?

Perhaps this is behaviour that we can toggle with an additional setting? Then we can leave it up to the end-user. I'm loathe to suggest adding an additional setting, but do want to put the option on the table.

Reply:
> What I meant is that languages such as PHP generate HTML files which are served with the .php extension.

Apologies for being pedantic, but I don't think this is quite correct, at least as far as I understand it. What happens when you want to run PHP files is the web server receives a request for a path and is configured to run a CGI script for that path. The CGI script reads the file at that path, runs it, and returns the result of the script in the body of the HTTP response to the client. This is usually HTML but it doesn't have to be in general (especially these days, it tends to be JSON which is read by the frontend). There's never an HTML file as such that you can get to on the server, it's always generated on the fly and returned in the response. The reason I'm pointing this out is I think it makes things a little more complicated for redirects from static sites where there _are_ actual HTML files present.

> I define in the web server that the MIME type text/html should be any file which ends in htm, html, shtml, php.

I suppose you could do that, but it just feels wrong to me to trick the browser into redirecting in this way 🤣 You might as well save the response and just configure the web server to redirect from the old paths, if you're having to configure the web server anyways.

> In other words, if I were to have PHP installed on my server, and I have the files a.php and b.php/index.html, would the web server be smart enough to know that http://example.com/a.php should serve a.php through PHP, and http://example.com/b.php is a directory and it should server b.php/index.html?

Simply having PHP installed won't cause any problems. You'd have to configure the server to serve files at particular paths with the CGI script. If you want to mix PHP and static files, you should teach your PHP engine how to do that. The browser doesn't care what the file extension is, in general, and there is a difference at the filesystem level between a directory and a file, so I don't think a server configured to serve only static files will be confused by `*.php/index.html`. 😊

Lastly, do links to paths on a PHP-based website usually include the extension? If you leave it off, what happens?

Reply:
You're right. I shouldn't have said "generates an HTML file". It is a response which (usually) contains HTML.

So when I go to a location such as http://example.com/page.php, that address maps to a file on the server's file system, say `/srv/www/mysite/page.php`. The web server then instructs the PHP engine to process that file. The PHP engine processes the file, delivers the response (HTML, plain text, JSON, or whatever), and includes it in the response to the browser's request to view the web address.

But that got me thinking. We were wrong in saying that Pelican doesn't generate PHP files. It can. It generates HTML without any PHP directives. So if Pelican were to generate a `.php` file (which only contains HTML), the PHP engine would look at contents of the file, not find any PHP directives `<?php ... ?>`, and thus return the page as-is.

A web server would typically need to know the extension of a PHP file in order to know to pass it on to the PHP engine, unless (perhaps) rules are explicitly defined to act otherwise. Another example: if http://example.com/some/page is a website written in PHP, and that address displays a page, it's not because `page` is actually `page.php` on the server, but because the page on the server is `some/page/index.php`. You can see this in a live demo of WordPress: https://s1.demo.opensourcecms.com/wordpress/wp-login.php  The extension `.php` is required; if we remove it, we get a 404 response.

Reply:
> Another example: if http://example.com/some/page is a website written in PHP, and that address displays a page, it's not because `page` is actually `page.php` on the server, but because the page on the server is `some/page/index.php`

So generating `index.html` that redirects to the new location would be fine in this case, right?

> The PHP engine processes the file, delivers the response (HTML, plain text, JSON, or whatever), and includes it in the response to the browser's request to view the web address.

So as far as I understand, generating a _directory_ `/srv/www/mysite/page.php/` whose contents are an `index.html` file that does a redirect would be fine. Is that correct?

> So if Pelican were to generate a `.php` file (which only contains HTML), the PHP engine would look at contents of the file, not find any PHP directives `<?php ... ?>`, and thus return the page as-is.

This is I think where we're talking past each other 😄 If you're using a PHP engine, why are you using Pelican at all? I mean, Pelican can technically generate any kind of file, it's "just" an interface to a templating engine that provides a lot of nice things with respect to managing a website... But if you have PHP on the server, and you're running a PHP process to receive requests from the web server, you don't need Pelican, right?

Reply:
> So generating some/page/index.html that redirects to the new location would be fine in this case, right?

💯

> So as far as I understand, generating a directory .../page.php/ whose contents are an index.html file that does a redirect would be fine. Is that correct?

I don't actually know. I would need to test. The behaviour might also depend on the specific web server being used. It all comes down to how the web server maps an address to the file system: whether it smart enough to know that a directory should be referenced, or whether it assumes that anything ending in `.php` must be a file (a legacy hyperlink won't include the trailing slash).

> This is I think where we're talking past each other 

Yeah, I guess me trying to explain stuff late at night isn't the best. My intention is simply that if someone is using this plugin, they are probably migrating from another site and want to keep old hyperlinks working. If they were previously using an CMS based on PHP (for example), there might be a bunch of files ending in `.php`. Their content is now going to be plain HTML, and we want to redirect (somehow) from the old address to the new one.  Your suggestion of putting an `index.html` in a directory ending in `.php` is the cleanest, provided that web servers play along.

Alternatively, I can incorporate your suggested changes into this PR, and we make the problem of how to handle redirects with funny file extensions a future problem. 🙃

Reply:
Incidentally just noticed this in the Pelican output:

```
[21:44:29] WARNING  Unable to find `/favicon.ico` or variations:
                    /favicon.ico.html                                                                                                                                                                                                       
                    /favicon.ico/index.html                                                                                                                                                                                                 
                    /favicon.ico
```
It does look in `/favicon.ico/index.html`, so it looks like Pelican itself assumes that a web server can resolve an address with an extension to a directory.

Reply:
> Your suggestion of putting an index.html in a directory ending in .php is the cleanest, provided that web servers play along.

I think that's as much as I want this extension to support, anyhow 😊 in which case, I think my suggested edit, with the change to add `.htm` should suffice here?